include/onnxruntime/core/framework/execution_provider.h

```diff
@@ -96,10 +98,13 @@ class IExecutionProvider {
      contains more than one node. The node indexes contained in sub-graphs may
      have overlap, and it's ONNXRuntime's responsibility to do the partition
      and decide whether a node will be assigned to <*this> execution provider.
+     `kernel_type_str_resolver` can be used to look up kernels in
+     `kernel_registries`.
   */
   virtual std::vector<std::unique_ptr<ComputeCapability>>
   GetCapability(const onnxruntime::GraphViewer& graph_viewer,
-                const std::vector<const KernelRegistry*>& kernel_registries) const;
+                const std::vector<const KernelRegistry*>& kernel_registries,
+                const IKernelTypeStrResolver& kernel_type_str_resolver) const;
```

Comment:
I think either of those options, a kernel lookup functor or removing `kernel_registries`, would be an improvement.

KernelRegistryManager::GetKernelRegistriesByProviderType() is what currently gets passed to GetCapability() and it can contain custom kernel registries. I'm not sure about how arbitrary EPs should handle custom kernel registries, or if they need to be handled.

Reply:
I don't quite understand how an EP would have an implementation for a custom kernel defined in a registry it didn't own. @pranavsharma do you know of any scenario where that would be possible?

`kernel_type_str_resolver` is also slightly confusing if we update the usage to not pass in kernel_registries, in that the EP owns the registry, but to do the lookup you're giving it a resolver that is external. Would schema_type_str_resolver be clearer?

Reply:
The name "kernel type string" is deliberate (but suggestions are welcome). It specifically refers to a type string used in kernel def type constraints, i.e., a type parameter like "T" or an input/output name.

This is different from type strings used in op schema type constraints, i.e., a type parameter like "T" or the actual type.

Reply:
Answering @skottmckay's question

> I don't quite understand how an EP would have an implementation for a custom kernel defined in a registry it didn't own. @pranavsharma do you know of any scenario where that would be possible?

A user is allowed to write custom ops [targeted towards a certain EP](  const char*(ORT_API_CALL* GetExecutionProviderType)(_In_ const struct OrtCustomOp* op);
) and register them with ORT. All these custom ops get accumulated in a custom op registry and get [registered](https://github.com/microsoft/onnxruntime/blob/0b0c51e02890ea35d0e5023681f2362421e44ceb/onnxruntime/core/session/inference_session.cc#L542-L547) in the inference session. Each kernel registry may contain kernels from many different providers. EPs don't own these registries; they're co-owned by the infer session and the users. This is the reason they're passed to the EPs in the GetCapability call. Now, I think only the custom registries should be passed to the EPs in the GetCapability call since the EPs already own the registries they created themselves, but that's a different story.